analysis/src/cdecl.rs

```diff
@@ -0,0 +1,478 @@
+    UnsupportedRightmostToken(CTok<'a>),
+    UnbalancedBrackets,
+    UnsupportedArrayLength,
+}
+
+impl<'a> CDecl<'a> {
+    // HACK(eddyb) this split is literally just to simplify error tracking.
+    pub fn parse<'b>(
+        mode: CDeclMode,
+        tokens: &'b [CTok<'a>],
+    ) -> Result<CDecl<'a>, CDeclParseError<'a, 'b>> {
+        CDecl::parse_inner(mode, tokens).map_err(|kind| CDeclParseError { kind, tokens })
+    }
+    fn parse_inner<'b>(
+        mode: CDeclMode,
+        tokens: &'b [CTok<'a>],
+    ) -> Result<CDecl<'a>, CDeclParseErrorKind<'a>> {
+        use CDeclParseErrorKind as ErrorKind;
+
+        trait InsertIfNone<T> {
+            fn insert_if_none(&mut self, value: T) -> Option<&mut T>;
+        }
+        impl<T> InsertIfNone<T> for Option<T> {
+            fn insert_if_none(&mut self, value: T) -> Option<&mut T> {
+                self.is_none().then(|| self.insert(value))
+            }
+        }
+
+        let (mut left, decl_name, mut right) = {
+            let mut decl_names =
+                tokens
+                    .iter()
+                    .copied()
+                    .enumerate()
+                    .filter_map(|(i, tok)| match tok {
+                        CTok::DeclName(name) => Some((i, name)),
+
+                        // HACK(eddyb) this is only allowed due to the (few)
+                        // function pointer typedefs in `vk.xml`, which don't
+                        // label parameter names in any special way.
+                        CTok::StrayIdent(name) if mode == CDeclMode::FuncTypeParam => {
+                            Some((i, name))
+                        }
+
+                        _ => None,
+                    });
+            match (decl_names.next(), decl_names.next()) {
+                (Some((i, name)), None) => (&tokens[..i], name, &tokens[i + 1..]),
+                (None, _) => return Err(ErrorKind::Missing("DeclName")),
+                (Some(_), Some(_)) => return Err(ErrorKind::Multiple("DeclName")),
+            }
+        };
+
+        if mode == CDeclMode::TypeDef {
+            // NOTE(eddyb) `typedef` can appear later on as well, so this is
+            // unnecessarily strict, but it avoids much more involved tracking.
+            left = left
+                .strip_prefix(&[CTok::Kw("typedef")])
+                .ok_or(ErrorKind::Missing("typedef"))?;
+            right = right
+                .strip_suffix(&[CTok::Punct(';')])
+                .ok_or(ErrorKind::Missing(";"))?;
+        }
+
+        let bitfield_width = match right {
+            [rest @ .., CTok::Punct(':'), CTok::IntLit(width_lit)]
+                if mode == CDeclMode::StructMember =>
+            {
+                right = rest;
+                Some(width_lit.parse().map_err(ErrorKind::InvalidIntLit)?)
+            }
+            _ => None,
+        };
+
+        // FIXME(eddyb) deduplicate qualifier parsing somehow.
+        let mut const_qualif = match left {
```

Comment:
Having this be an `Option<()>` kind of feels unnecessary to me. It could just as well be a `bool`.

```diff
diff --git a/analysis/src/cdecl.rs b/analysis/src/cdecl.rs
index 105ef89..dcf4522 100644
--- a/analysis/src/cdecl.rs
+++ b/analysis/src/cdecl.rs
@@ -1,4 +1,4 @@
-use std::num::NonZeroU8;
+use std::{mem, num::NonZeroU8};
 
 /// Identifier-category-aware minimal tokenization of a subset of C syntax,
 /// sufficient for parsing the C declarations used in `vk.xml`.
@@ -155,15 +155,6 @@ impl<'a> CDecl<'a> {
     ) -> Result<CDecl<'a>, CDeclParseErrorKind<'a>> {
         use CDeclParseErrorKind as ErrorKind;
 
-        trait InsertIfNone<T> {
-            fn insert_if_none(&mut self, value: T) -> Option<&mut T>;
-        }
-        impl<T> InsertIfNone<T> for Option<T> {
-            fn insert_if_none(&mut self, value: T) -> Option<&mut T> {
-                self.is_none().then(|| self.insert(value))
-            }
-        }
-
         let (mut left, decl_name, mut right) = {
             let mut decl_names =
                 tokens
@@ -214,9 +205,9 @@ impl<'a> CDecl<'a> {
         let mut const_qualif = match left {
             [CTok::Kw("const"), rest @ ..] => {
                 left = rest;
-                Some(())
+                true
             }
-            _ => None,
+            _ => false,
         };
 
         let mut ty = CType::Base(match left {
@@ -262,14 +253,16 @@ impl<'a> CDecl<'a> {
             while let Some((&leftmost, after_leftmost)) = left.split_first() {
                 match leftmost {
                     CTok::Kw("const") => {
-                        const_qualif
-                            .insert_if_none(())
-                            .ok_or(ErrorKind::Multiple("const"))?;
+                        if const_qualif {
+                            return Err(ErrorKind::Multiple("const"));
+                        }
+
+                        const_qualif = true;
                     }
                     CTok::Punct('*') => {
                         ty = CType::Ptr {
                             implicit_for_decay: false,
-                            is_const: const_qualif.take().is_some(),
+                            is_const: mem::take(&mut const_qualif),
                             pointee: Box::new(ty),
                         };
                     }
@@ -336,7 +329,7 @@ impl<'a> CDecl<'a> {
                         };
                     }
                     [CTok::Punct('('), params @ .., CTok::Punct(')')] => {
-                        if const_qualif.is_some() {
+                        if const_qualif {
                             return Err(ErrorKind::Unused("const"));
                         }
 
@@ -375,12 +368,12 @@ impl<'a> CDecl<'a> {
         if let (CDeclMode::FuncParam, CType::Array { .. }) = (mode, &ty) {
             ty = CType::Ptr {
                 implicit_for_decay: true,
-                is_const: const_qualif.take().is_some(),
+                is_const: mem::take(&mut const_qualif),
                 pointee: Box::new(ty),
             };
         }
 
-        if const_qualif.is_some() {
+        if const_qualif {
             return Err(ErrorKind::Unused("const"));
         }
 ```

Reply:
I had more uses for `insert_if_none` and there's a lot more things in the C syntax that work similarly, but are not used in here, I'll have to take a second look.

Booleans aren't great for stuff like this for a few different reasons, but I'm almost certain `insert_if_none` had more call sites before I refactored all of this to be more general and work for the function pointer typedefs, so I should clean it up.